ngraph/core/include/ngraph/runtime/reference/interpolate.hpp

```diff
@@ -0,0 +1,593 @@
+                                                      height_scale,
+                                                      static_cast<float>(output_height),
+                                                      static_cast<float>(input_height));
+                    y_original.push_back(in_y);
+                    in_y = std::max(0.0f, std::min(in_y, static_cast<float>(input_height - 1)));
+
+                    const int64_t in_y1 = std::min(static_cast<int64_t>(in_y), input_height - 1);
+                    const int64_t in_y2 = std::min(in_y1 + 1, input_height - 1);
+                    dy1[y] = std::fabs(in_y - in_y1);
+                    dy2[y] = std::fabs(in_y - in_y2);
+
+                    if (in_y1 == in_y2)
+                    {
+                        dy1[y] = 0.5f;
+                        dy2[y] = 0.5f;
+                    }
+
+                    input_width_mul_y1[y] = input_width * in_y1;
+                    input_width_mul_y2[y] = input_width * in_y2;
+                }
+
+                for (int64_t x = 0; x < output_width; ++x)
+                {
+                    float in_x = m_get_original_coord(static_cast<float>(x),
+                                                      width_scale,
+                                                      static_cast<float>(output_width),
+                                                      static_cast<float>(input_width));
+                    x_original.push_back(in_x);
+                    in_x = std::max(0.0f, std::min(in_x, static_cast<float>(input_width - 1)));
+
+                    in_x1[x] = std::min(static_cast<int64_t>(in_x), input_width - 1);
+                    in_x2[x] = std::min(in_x1[x] + 1, input_width - 1);
+
+                    dx1[x] = std::abs(in_x - in_x1[x]);
+                    dx2[x] = std::abs(in_x - in_x2[x]);
+                    if (in_x1[x] == in_x2[x])
+                    {
+                        dx1[x] = 0.5f;
+                        dx2[x] = 0.5f;
+                    }
+                }
+
+                const T* xdata = input_data;
+                T* ydata = out;
+                for (int64_t n = 0; n < batch_size; ++n)
+                {
+                    for (int64_t c = 0; c < num_channels; ++c)
+                    {
+                        for (int64_t y = 0; y < output_height; ++y)
+                        {
+                            for (int64_t x = 0; x < output_width; ++x)
+                            {
+                                T x11 = xdata[input_width_mul_y1[y] + in_x1[x]];
+                                T x21 = xdata[input_width_mul_y1[y] + in_x2[x]];
+                                T x12 = xdata[input_width_mul_y2[y] + in_x1[x]];
+                                T x22 = xdata[input_width_mul_y2[y] + in_x2[x]];
+
+                                ydata[output_width * y + x] =
+                                    static_cast<T>(dx2[x] * dy2[y] * x11 + dx1[x] * dy2[y] * x21 +
```

Comment:
Looks like only this code requires template argument.

Please try to minimize template code. And measure the impact on library size.

Reply:
No, it is not true: template parameter is used in all modes of Interpolate-4.

Reply:
But you can decrease the number of template code. Currently you create huge template functions (100+ lines of code) but template argument needed only for several lines. Please separate the template code.

It should help you to decrease impact on binary size.

Reply:
Size of this method was reduced: all non-template parts is in helper class.